scapy/contrib/automotive/gm/gmlanutils.py

```diff
@@ -0,0 +1,384 @@
+# See http://www.secdev.org/projects/scapy for more information
+# Copyright (C) Markus Schroetter <project.m.schroetter@gmail.com>
+# Copyright (C) Nils Weiss <nils@we155.de>
+# This program is published under a GPLv2 license
+
+# scapy.contrib.description = GMLAN Utilities
+# scapy.contrib.status = loads
+
+import time
+from scapy.contrib.automotive.gm.gmlan import GMLAN, GMLAN_SA, GMLAN_RD, \
+    GMLAN_TD, GMLAN_PM, GMLAN_RMBA
+from scapy.config import conf
+from scapy.contrib.isotp import ISOTPSocket
+from scapy.error import warning, log_loading
+from scapy.utils import PeriodicSenderThread
+
+
+__all__ = ["GMLAN_TesterPresentSender", "GMLAN_InitDiagnostics",
+           "GMLAN_GetSecurityAccess", "GMLAN_RequestDownload",
+           "GMLAN_TransferData", "GMLAN_TransferPayload",
+           "GMLAN_ReadMemoryByAddress", "GMLAN_BroadcastSocket"]
+
+log_loading.info("\"conf.contribs['GMLAN']"
+                 "['treat-response-pending-as-answer']\" set to True). This "
+                 "is required by the GMLAN-Utils module to operate "
+                 "correctly.")
+try:
+    conf.contribs['GMLAN']['treat-response-pending-as-answer'] = False
+except KeyError:
+    conf.contribs['GMLAN'] = {'treat-response-pending-as-answer': False}
+
+
+class GMLAN_TesterPresentSender(PeriodicSenderThread):
+    def __init__(self, sock, pkt=GMLAN(service="TesterPresent"), interval=2):
+        """ Thread to send TesterPresent messages packets periodically
+
+        Args:
+            sock: socket where packet is sent periodically
+            pkt: packet to send
+            interval: interval between two packets
+        """
+        PeriodicSenderThread.__init__(self, sock, pkt, interval)
+
+
+def GMLAN_InitDiagnostics(sock, broadcastsocket=None, timeout=None,
+                          verbose=None, retry=0):
+    """Send messages to put an ECU into an diagnostic/programming state.
+
+    Args:
+        sock:       socket to send the message on.
+        broadcast:  socket for broadcasting. If provided some message will be
+                    sent as broadcast. Recommended when used on a network with
+                    several ECUs.
+        timeout:    timeout for sending, receiving or sniffing packages.
+        verbose:    set verbosity level
+        retry:      number of retries in case of failure.
+
+    Returns true on success.
+    """
+    if verbose is None:
+        verbose = conf.verb
+    retry = abs(retry)
+
+    while retry >= 0:
+        retry -= 1
+
+        # DisableNormalCommunication
+        p = GMLAN(service="DisableNormalCommunication")
+        if broadcastsocket is None:
+            if verbose:
+                print("Sending %s" % repr(p))
+            resp = sock.sr1(p, timeout=timeout, verbose=0)
+            if resp is not None:
+                if verbose:
+                    resp.show()
+                if resp.sprintf("%GMLAN.service%") != "DisableNormalCommunicationPositiveResponse":   # noqa: E501
```

Comment:
`resp.sprintf("%GMLAN.service%") == "NegativeResponse"` would be more natural in my opinion

Reply:
While it's true that there should only be this exact positive response or the negative response, i think it would be more robust. The code should only proceed to the next step when everything succeeded - which `!= DisableNormalCommunicationPositiveResponse` would express better. What do you think?

Reply:
well, the only difference would be, you can have the identical code in all cases.

Reply:
and, you are testing for a negative condition, because you want to continue if ...